### ISSUE TITLE
Unable to compile under OpenBSD/amd64 [resolved]

### ISSUE THREAD

**Issue by heinrichrebehn** [mikebrady/shairport-sync#658]
Hi all,
has anybody been able to compile shairport-sync under OpenBSD/amd64? I downloaded the source and needed utilities, but when I issue 'autoreconf -i -f' the command runs forever eating up > 90% CPU.

Any ideas?

-Heinrich

**Comment by mikebrady** [comment]
Hi there. Have you looked at the FREEBSD page? (Sorry if it's an obvious question...)

**Comment by mikebrady** [comment]
Also -- are you compiling the regular (`master`) branch or the `development` branch?

**Comment by heinrichrebehn** [comment]
I saw this mail only now. No, I have not looked yet. Will do so now..

-Heinrich

**Comment by heinrichrebehn** [comment]
Hi Mike,

Thank you for your quick reply!

I am trying to compile from shairport-sync-3.1.7.tar.gz
The problem that i initially reported, has been solved. It was an improper installation of autoconf.

Next problem was that OpenBSD does not have librt. I manually removed the related lines from configure.ac

Next issue was: Although i specified —without-alsa to ./configure I got errors saying ALSA was not found. Omitting this flag cleared the error.

Now "./configure --with-sndio --with-avahi” completes without error, but when I issue “make”, I get

shairport.c:31:23: error: libconfig.h: No such file or directory

# find / -name libconfig.h
/usr/local/include/libconfig.h

Using ktrace(1) I found that libconfig.h is only looked for in /usr/include, not in /usr/local/include.

I linked /usr/local/include/libconfig.h to /usr/include/

Now I get:

make  all-recursive
Making all in man
depbase=`echo shairport.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`; gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/usr/local/etc\"   -Wno-multichar -DSYSCONFDIR=\"/usr/local/etc\"   -g -O2 -MT shairport.o -MD -MP -MF $depbase.Tpo -c -o shairport.o shairport.c && mv -f $depbase.Tpo $depbase.Po
In file included from shairport.c:34:
/usr/include/net/if.h:121: error: field 'ifi_lastchange' has incomplete type
/usr/include/net/if.h:359: error: field 'ifru_addr' has incomplete type
/usr/include/net/if.h:360: error: field 'ifru_dstaddr' has incomplete type
/usr/include/net/if.h:361: error: field 'ifru_broadaddr' has incomplete type
/usr/include/net/if.h:388: error: field 'ifrau_addr' has incomplete type
/usr/include/net/if.h:394: error: field 'ifra_dstaddr' has incomplete type
/usr/include/net/if.h:396: error: field 'ifra_mask' has incomplete type
/usr/include/net/if.h:439: error: field 'addr' has incomplete type
/usr/include/net/if.h:440: error: field 'dstaddr' has incomplete type
/usr/include/net/if.h:446: error: expected specifier-qualifier-list before 'sa_family_t'
In file included from /usr/include/net/if.h:455,
                 from shairport.c:34:
/usr/include/net/if_arp.h:79: error: field 'arp_pa' has incomplete type
/usr/include/net/if_arp.h:80: error: field 'arp_ha' has incomplete type
shairport.c:35:18: error: popt.h: No such file or directory
shairport.c:78:29: error: libdaemon/dexec.h: No such file or directory

.. and many more

I am at a loss now.

Greetings,

	Heinrich Rebehn

**Comment by mikebrady** [comment]
Thanks. Could you list the packages you've installed so far?

**Comment by heinrichrebehn** [comment]
# pkg_info 
autoconf-2.69p2     automatically configure source code on many Un*x platforms
automake-1.15.1     GNU Standards-compliant Makefile generator
avahi-0.7p0         framework for Multicast DNS Service Discovery
bash-4.4.12p0       GNU Bourne Again Shell
bzip2-1.0.6p8       block-sorting file compressor, unencumbered
dbus-1.10.22p0v0    message bus system
dbus-daemon-launch-helper-1.10.22p0 DBus setuid helper for starting system services
gdbm-1.13           GNU dbm
gdiff-3.6           GNU versions of the diff utilities
gettext-0.19.8.1p1  GNU gettext runtime libraries and programs
glib2-2.52.3        general-purpose utility library
gmake-4.2.1         GNU make
libconfig-1.5p1     library for manipulating structured configuration files
libdaemon-0.14p1    lightweight C library that eases the writing of daemons
libelf-0.8.13p3     read, modify, create ELF files on any arch
libffi-3.2.1p2      Foreign Function Interface
libiconv-1.14p3     character set conversion library
libsigsegv-2.10p2   library for handling page faults in user mode
libslang-2.2.4p2    stack-based interpreter for terminal applications
libssh2-1.8.0       library implementing the SSH2 protocol
mc-4.8.19           free Norton Commander clone with many useful features
metaauto-1.0p1      wrapper for gnu auto*
oniguruma-5.9.6     regular expressions library
pcre-8.40p1         perl-compatible regular expression library
png-1.6.31          library for manipulating PNG images
popt-1.16p1         getopt(3)-like library with a number of enhancements
python-2.7.14       interpreted object-oriented programming language
quirks-2.367        exceptions to pkg_add rules
sqlite3-3.20.1      embedded SQL implementation
unzip-6.0p11        extract, list & test files in a ZIP archive
zip-3.0p0           create/update ZIP files compatible with PKZip(tm)

**Comment by mikebrady** [comment]
Thanks. I'll try to take a log at it over the next few days.

**Comment by heinrichrebehn** [comment]
Thank you for your kind support!

-Heinrich

**Comment by mikebrady** [comment]
Hi there. I think I got it working, but I'm not quite sure how, so I'm going to try to reproduce the steps and get back to you. It looks like there might be a silent bug in Linux and FreeBSD in `configure.ac`.

Just to tantalise --  sorry, can't resist :) 
```
$ ./shairport-sync -V
3.1.7-OpenSSL-Avahi-sndio-sysconfdir:/usr/local/etc
$ uname -a
OpenBSD openbsd.localdomain 6.2 GENERIC.MP#134 amd64
$ 
```

**Comment by heinrichrebehn** [comment]
Great! I hope that it will compile on arm too. I intend to buy a Raspberry Pi 3 and use it as an airplay receiver under OpenBSD.
Thank you so much for your efforts!

-Heinrich

**Comment by mikebrady** [comment]
Thanks. It’s not clear yet that it’s fully working... it will need a little more investigation.

**Comment by mikebrady** [comment]
There is now a very preliminary [OPENBSD.md](https://github.com/mikebrady/shairport-sync/blob/master/OPENBSD.md) page. I'd be grateful if you could test it and report any problems.

I have not been able to test this -- I know it compiles and it seems to generate audio, but I'm running it in an installation with no sound hardware, so I can't test it. Maybe you can.

**Comment by heinrichrebehn** [comment]
Thanks! I think I will be able to test ist over the weekend and will report back to you.

-Heinrich

**Comment by heinrichrebehn** [comment]
Hi Mike,

I tested with OpenBSD 6.2 running virtualised under VMware Fusion with sound output via a cheap USB dongle.

Your development branch compiles and runs fine under both virtual amd64 and i386 machines.

Music was played using iTunes on a Mac and sent simultaneously to an Airport Express and shairport-sync. Both audio streams were perfectly in sync! No slapback echo, no fake stereo (doubling).

Thank you so much for your work! Next step for me will be purchasing a Raspberry Pi 3 and run shairport-sync there. Can I expect to compile as smooth as on intel? I have absolutely no experience with other platforms, but autoconf/make should work just the same?

With best wishes,

	Heinrich

**Comment by mikebrady** [comment]
That's great news, Heinrich! 

My OpenBSD 6.2 is also in VmWare Fusion -- can you tell me what you did to get the USB DAC connected please? Did you have to install VMWare Tools and what audio settings did you need to set.

I'm afraid I have really no knowledge of how OpenBSD runs on a Pi, but if it's reasonably complete, Shairport Sync should compile and run -- autoconf is indeed great.

Please keep us all posted.

**Comment by heinrichrebehn** [comment]
It was very simple, no magic involved. No VMWare Tools, i was not even aware that they exist for OpenBSD.
I
	- removed sound card from virtual machine settings
	- booted machine
	- plugged in USB DAC
	- VMware asks wether device should be connected to MacOS or virtual machine
	- answer question appropriately and see:

uaudio0 at uhub2 port 1 configuration 1 interface 0 "ABC C-Media USB Audio Device" rev 1.10/1.00 addr 4
uaudio0: audio rev 1.00, 8 mixer controls
audio0 at uaudio0
uhidev2 at uhub2 port 1 configuration 1 interface 3 "ABC C-Media USB Audio Device" rev 1.10/1.00 addr 4
uhidev2: iclass 3/0
uhid0 at uhidev2: input=4, output=4, feature=0

	on the virtual machine’s console.

sndio connects to the new device and you can start shairport-sync.

However, sound support via USB DAC is quite flakey. It works fine while you leave the host machine alone, but as soon as you open a browser window or similar, sound stops. This is not related to shairport, it also happens when you play an audio file using sox(1).

> On 10. Feb 2018, at 19:46, Mike Brady <notifications@github.com> wrote:
> 
> That's great news, Heinrich!
> 
> My OpenBSD 6.2 is also in VmWare Fusion -- can you tell me what you did to get the USB DAC connected please? Did you have to install VMWare Tools and what audio settings did you need to set.
> 
> I'm afraid I have really no knowledge of how OpenBSD runs on a Pi, but if it's reasonably complete, Shairport Sync should compile and run -- autoconf is indeed great.
> 
> Please keep us all posted.

It will probably take a couple of weeks to get and install the Raspi, but I will either report success or bother you with questions ;-)

-Heinrich

**Comment by mikebrady** [comment]
Many thanks.

**Comment by heinrichrebehn** [comment]
Hi Mike,

I just wanted to get back to you and report about shairport-sync on Raspi under OpenBSD.
The good news: it compiles fine :-) It can also be started and runs fine.

The bad news: Onboard sound is not supported on the Raspi and any attempt to access a USB audio device results in a kernel panic:

https://marc.info/?l=openbsd-bugs&m=151897155509345&w=2

While waiting for a fix I installed Raspbian with shairport-sync via apt-get install. It works OOB without any delays!

Thank you for your good work!

P.S.

I also tried to install under macOS High Sierra using MacPorts. It installs:

Version: "2.8.6-OpenSSL-ao-stdout-pipe-soxr-metadata-sysconfdir:/opt/local/etc/shairport-sync”
Unfortunately it produces a horrible delay (almost a second)

I then tried:

git clone https://github.com/mikebrady/shairport-sync.git
Cloning into 'shairport-sync'...
remote: Counting objects: 8700, done.
remote: Compressing objects: 100% (93/93), done.
remote: Total 8700 (delta 67), reused 57 (delta 26), pack-reused 8581
Receiving objects: 100% (8700/8700), 4.15 MiB | 2.12 MiB/s, done.
Resolving deltas: 100% (5709/5709), done.

cd shairport-sync/
heinrich@host35 [~/shairport-sync]: autoconf -i -f
configure.ac:6: error: possibly undefined macro: AM_INIT_AUTOMAKE
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
configure.ac:28: error: possibly undefined macro: AC_MSG_ERROR
configure.ac:35: error: possibly undefined macro: AM_CONDITIONAL
configure.ac:64: error: possibly undefined macro: AC_DEFINE
heinrich@host35 [~/shairport-sync]: 

Any thoughts about this?

I am totally happy that it runs on Raspi, but being able to use my PC as an Airplay Server would be even better :-)

-Heinrich
